community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureRegistry.java

```diff
@@ -17,48 +17,74 @@
 
 import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
 import java.util.Map;
-import java.util.stream.Stream;
+import java.util.Set;
 
+import org.neo4j.collection.RawIterator;
 import org.neo4j.kernel.api.exceptions.ProcedureException;
 import org.neo4j.kernel.api.exceptions.Status;
+import org.neo4j.kernel.api.proc.Procedure;
+import org.neo4j.kernel.api.proc.ProcedureSignature;
 
-
-public class Procedures
+public class ProcedureRegistry
 {
     private final Map<ProcedureSignature.ProcedureName,Procedure> procedures = new HashMap<>();
 
     /**
      * Register a new procedure. This method must not be called concurrently with {@link #get(ProcedureSignature.ProcedureName)}.
+     *
      * @param proc the procedure.
      */
     public synchronized void register( Procedure proc ) throws ProcedureException
     {
         ProcedureSignature signature = proc.signature();
         ProcedureSignature.ProcedureName name = signature.name();
-        if( procedures.putIfAbsent( name, proc ) != null )
+
+        validateSignature( signature, signature.inputSignature(), "input" );
+        validateSignature( signature, signature.outputSignature(), "output" );
+
+        if ( procedures.putIfAbsent( name, proc ) != null )
+        {
+            throw new ProcedureException( Status.Procedure.FailedRegistration,
+                    "Unable to register procedure, because the name `%s` is already in use.", name );
+        }
+    }
+
+    private void validateSignature( ProcedureSignature signature, List<ProcedureSignature.FieldSignature> fields, String fieldType )
+            throws ProcedureException
+    {
+        Set<String> names = new HashSet<>();
+        for ( ProcedureSignature.FieldSignature field : fields )
         {
-            throw new ProcedureException( Status.Procedure.FailedRegistration, "Unable to register procedure, because the name `%s` is already in use.", name );
+            if ( !names.add( field.name() ) )
+            {
+                throw new ProcedureException( Status.Procedure.FailedRegistration,
+                        "Procedure `%s` cannot be registered, because it contains a duplicated " + fieldType + " field, '%s'. " +
+                        "You need to rename or remove one of the duplicate fields.", signature.toString(), field.name() );
+            }
         }
     }
 
     public ProcedureSignature get( ProcedureSignature.ProcedureName name ) throws ProcedureException
     {
         Procedure proc = procedures.get( name );
-        if( proc == null )
+        if ( proc == null )
         {
             throw noSuchProcedure( name );
         }
         return proc.signature();
     }
 
-    public Stream<Object[]> call( Procedure.Context ctx, ProcedureSignature.ProcedureName name, Object[] input ) throws ProcedureException
+    public RawIterator<Object[],ProcedureException> call( Procedure.Context ctx, ProcedureSignature.ProcedureName name, Object[] input )
```

Comment:
really strange concurrency semantics - call and get not synchronized in the same time all register methods are, looks like potential source of issues